botan3: init at 3.1.1

thillux opened the pull request:
###### Description of changes

botan was released in version 3 earlier this year. Version 3 most importantly adds TLS 1.3 and support for PQC algorithms. Reuse existing generic botan file. Drop patching of C++ standard, which was added for botan 2.0.1 some time ago and is no longer needed.

See: https://botan.randombit.net/news.html

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

thillux committed:
botan3: init at 3.1.1

Use the existing generic botan file and add specialization
for botan 3. Botan 3 most importantly adds support for
TLS 1.3 and PQC algorithms. Introduce Botan 3 in parallel to
Botan 2, as it is a major release and e.g. now uses C++20
in contrast to C++11 of Botan 2.9.

Signed-off-by: Markus Theil <theil.markus@gmail.com>

thillux committed:
botan2: small cleanup

The explicit setting of the C++ standard to C++11 was
introduced with botan 2.0.1 and is no longer needed.

Signed-off-by: Markus Theil <theil.markus@gmail.com>

thillux commented:
Result of `nixpkgs-review pr 244212` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>biboumi</li>
    <li>botan2</li>
    <li>botan2.dev</li>
    <li>botan3</li>
    <li>botan3.dev</li>
    <li>chickenPackages_5.chickenEggs.botan</li>
    <li>corectrl</li>
    <li>ikiwiki-full</li>
    <li>keepassxc</li>
    <li>monotone</li>
    <li>qownnotes</li>
    <li>rnp</li>
    <li>rnp.dev</li>
    <li>rnp.lib</li>
    <li>softhsm</li>
    <li>step-kms-plugin</li>
  </ul>
</details>

nikstur commented on pkgs/development/libraries/botan/3.0.nix:
```diff
@@ -0,0 +1,7 @@
+callPackage ./generic.nix (args // {
+  baseVersion = "3.1";
+  revision = "1";
+  sha256 = "sha256-MMhP6RmTapj+9TMfJGxiqiwOTSCFstRREgf2ogr6Oms=";
```
```suggestion
  meta.badPlatforms = [ lib.platforms.darwin ];
```

nikstur commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20292,6 +20292,10 @@ with pkgs;
     inherit (darwin.apple_sdk.frameworks) CoreServices Security;
   };
 
+  botan3 = callPackage ../development/libraries/botan/3.0.nix {
+    inherit (darwin.apple_sdk.frameworks) CoreServices Security;
+  };
```
```suggestion
  botan3 = callPackage ../development/libraries/botan/3.0.nix {  };
```

nikstur reviewed (changes requested):
Looks like botan3 doesn't build on darwin (macos) anymore: https://logs.ofborg.org/?key=nixos/nixpkgs.244212&attempt_id=20961051-b956-4926-975b-1ae6ce2a694c

I don't think that it fundamentally doesn't so lets add darwin to `meta.badPlatforms`. Unless you have a Mac and want to try to fix it of course.

thillux committed:
botan3: flag macos as bad platform for now

Default clang 11 on MacOS is too old for botan 3,
which requires at least clang 14.

Signed-off-by: Markus Theil <theil.markus@gmail.com>

thillux commented:
> Looks like botan3 doesn't build on darwin (macos) anymore: https://logs.ofborg.org/?key=nixos/nixpkgs.244212&attempt_id=20961051-b956-4926-975b-1ae6ce2a694c
> 
> I don't think that it fundamentally doesn't so lets add darwin to `meta.badPlatforms`. Unless you have a Mac and want to try to fix it of course.

Botan 3 raised the bar on its dependencies quite a bit. It seems, that the default Clang 11.0 on MacOS is just too old for C++20. As long as botan2 is used by most tools, I would mark MacOS as bad platform and add a comment to reconsider this after MacOS gets a more recent compiler.

thillux committed:
botan: add thillux as maintainer

Signed-off-by: Markus Theil <theil.markus@gmail.com>

nikstur reviewed (approved):
LGTM, conditional on ofborg succeeding.

7c6f434c commented:
@ofborg eval

nikstur commented on pkgs/development/libraries/botan/3.0.nix:
```diff
@@ -0,0 +1,9 @@
+  revision = "1";
+  sha256 = "sha256-MMhP6RmTapj+9TMfJGxiqiwOTSCFstRREgf2ogr6Oms=";
+  # reconsider removing this platform marking, when MacOS uses Clang 14.0+ by default.
+  meta.badPlatforms = [ lib.platforms.darwin ];
```
Ahh you would need to `overrrideAttrs` this derivation and add meta.badPlatforms there. My bad. Or you could add a `batPlatforms` flag to generic.nix

thillux replied:
I added a badPlatforms Flag.
